File: packages/python/qsu/file/getFileInfo.py
import os
import math
import stat as statModule
from .getFileExtension import getFileExtension
from .getFileName import getFileName
from ..format.fileSizeFormat import fileSizeFormat


def getFileInfo(filePath: str) -> dict:
	def dateToUnixTime(seconds: float) -> int:
		return math.floor(seconds)

	# The OSError is raised as it is. Re-raising `Exception(str(err))` dropped
	# `errno`, `strerror` and `filename`, so a caller could not tell a missing
	# file from a permission error, and the original traceback went with them.
	fileItem = os.stat(filePath)

	return {
		'success': True,
		# Read the mode already returned by `stat` instead of asking the
		# filesystem a second time through `os.path.isdir`.
		'isDirectory': statModule.S_ISDIR(fileItem.st_mode),
		'ext': getFileExtension(filePath),
		'size': fileItem.st_size,
		'sizeHumanized': fileSizeFormat(fileItem.st_size),
		'name': getFileName(filePath),
		# `os.path` follows the host platform, so a Windows path separates on
		# '\\' here as it does in the JavaScript implementation.
		'dirname': os.path.dirname(filePath),
		'path': os.path.abspath(filePath),
		'created': dateToUnixTime(fileItem.st_ctime),
		'modified': dateToUnixTime(fileItem.st_mtime),
	}

File: packages/javascript/lib/node/file/getFileInfo.ts
import { Stats } from 'node:fs';
import { stat } from 'node:fs/promises';
import { dirname, resolve as pathResolve } from 'node:path';
import { getFileExtension } from './getFileExtension.js';
import { fileSizeFormat } from '../../format/fileSizeFormat.js';
import { getFileName } from './getFileName.js';
import { FileInfo } from '../../_types/global';

export async function getFileInfo(filePath: string): Promise<FileInfo> {
	const dateToUnixTime = (date: Date): number => Math.floor(new Date(date).getTime() / 1000);

	// The filesystem error is thrown as it is. Re-throwing `new Error(message)`
	// dropped `code`, `errno` and `path`, so a caller could not tell ENOENT from
	// EACCES, and the original stack was lost with them.
	const fileItem: Stats = await stat(filePath);

	return {
		success: true,
		isDirectory: fileItem.isDirectory(),
		ext: getFileExtension(filePath),
		size: fileItem.size,
		sizeHumanized: fileSizeFormat(fileItem.size),
		name: getFileName(filePath),
		dirname: dirname(filePath),
		path: pathResolve(filePath),
		created: dateToUnixTime(fileItem.ctime),
		modified: dateToUnixTime(fileItem.mtime)
	};
}
